# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/promise/Promise.cpp
# Commit: 0bd696ed34ea
# Full Hash: 0bd696ed34eace2844a540509a5022dc57b48508
# Author: Zibi Braniecki <zbraniecki@mozilla.com>
# Date: 2019-03-27 04:44:00
# Regressor Bug: 1483036
# File Overlap Count: 2
# Description:
#   Bug 1483036 - Add MaybeResolveWithClone and MaybeRejectWithClone to Promise. r=smaug
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D24385
# ==============================================================================

diff -r b7663bdb3f6e -r 0bd696ed34ea dom/promise/Promise.cpp
--- a/dom/promise/Promise.cpp	Tue Mar 26 19:28:15 2019 +0000
+++ b/dom/promise/Promise.cpp	Tue Mar 26 19:34:29 2019 +0000
@@ -41,6 +41,7 @@
 #include "PromiseWorkerProxy.h"
 #include "WrapperFactory.h"
 #include "xpcpublic.h"
+#include "xpcprivate.h"
 
 namespace mozilla {
 namespace dom {
@@ -548,6 +549,32 @@
   }
 }
 
+void Promise::MaybeResolveWithClone(JSContext* aCx,
+                                    JS::Handle<JS::Value> aValue) {
+  JS::Rooted<JSObject*> sourceScope(aCx, JS::CurrentGlobalOrNull(aCx));
+  AutoEntryScript aes(GetParentObject(), "Promise resolution");
+  JSContext* cx = aes.cx();
+  JS::Rooted<JS::Value> value(cx, aValue);
+
+  xpc::StackScopedCloneOptions options;
+  options.wrapReflectors = true;
+  StackScopedClone(cx, options, sourceScope, &value);
+  MaybeResolve(aCx, value);
+}
+
+void Promise::MaybeRejectWithClone(JSContext* aCx,
+                                   JS::Handle<JS::Value> aValue) {
+  JS::Rooted<JSObject*> sourceScope(aCx, JS::CurrentGlobalOrNull(aCx));
+  AutoEntryScript aes(GetParentObject(), "Promise rejection");
+  JSContext* cx = aes.cx();
+  JS::Rooted<JS::Value> value(cx, aValue);
+
+  xpc::StackScopedCloneOptions options;
+  options.wrapReflectors = true;
+  StackScopedClone(cx, options, sourceScope, &value);
+  MaybeReject(aCx, value);
+}
+
 JSObject* Promise::GlobalJSObject() const {
   return mGlobal->GetGlobalJSObject();
 }